File: client-zig/src/service_caller.zig
const std = @import("std");
const protocol = @import("protocol");
const print = std.debug.print;
const net = std.Io.net;
const mad = protocol.mad;
const Backoff = @import("backoff.zig").Backoff;

pub const ServiceCallError = error{
    PayloadTypeMismatch,
    CallFailed,
};

// Generic RPC caller: connects to a Service's fixed convention socket path,
// runs the same key/value type-fingerprint handshake, then sends requests
// and decodes responses one at a time. call() is mutex-guarded so concurrent
// callers on the same ServiceCaller can't interleave requests/responses on
// the single underlying connection — mirrors what spine-go's internal
// request channel + single run() goroutine achieves, just via a lock instead
// of a queue.
pub fn ServiceCaller(comptime K: type, comptime V: type) type {
    comptime {
        if (mad.getRequiredSize(K) > protocol.globals.MAX_PACKET_SIZE) {
            @compileError("key type too big for globals.MAX_PACKET_SIZE");
        }
        if (mad.getRequiredSize(V) + 1 > protocol.globals.MAX_PACKET_SIZE) {
            @compileError("value type too big for globals.MAX_PACKET_SIZE");
        }
    }

    return struct {
        io: std.Io,
        namespace: []const u8,
        name: []const u8,
        conn: net.Stream,
        is_connected: bool = false,

        r_buf: [protocol.globals.MAX_PACKET_SIZE]u8 = undefined,
        w_buf: [protocol.globals.MAX_PACKET_SIZE]u8 = undefined,
        reader: net.Stream.Reader = undefined,
        writer: net.Stream.Writer = undefined,
        lock: std.Io.Mutex = .init,

        const Self = @This();

        // Same backoff-retry reasoning as Subscriber.connect: a type
        // mismatch is permanent (K/V are fixed at compile time), so it
        // returns immediately instead of retrying forever.
        pub fn connect(self: *Self, io: std.Io, namespace: []const u8, name: []const u8) !void {
            var backoff: Backoff = .{};

            while (true) {
                self.dial(io, namespace, name) catch |err| {
                    if (err == ServiceCallError.PayloadTypeMismatch) return err;

                    print("spine: failed to connect to service '{s}' ({any}), retrying in {d}ms\n", .{ name, err, backoff.ms });
                    try backoff.sleep(io);
                    continue;
                };
                return;
            }
        }

        // BUGFIX: this used to be a blanket `self.* = .{...}` literal that
        // included `.lock = .init`. That's fine for the very first connect
        // (self is freshly allocated, uninitialized memory), but dial() is
        // also called to reconnect from *inside* call() when is_connected is
        // false — and call() has already locked self.lock at that point.
        // Resetting it mid-call replaced the held lock with a fresh
        // *unlocked* one, so call()'s deferred unlock later found nothing
        // locked and Io.Mutex correctly panicked rather than silently
        // ignoring it. lock is a one-time thing, set by Node.newServiceCaller
        // before the first connect() — dial() must never touch it.
        fn dial(self: *Self, io: std.Io, namespace: []const u8, name: []const u8) !void {
            var path_buf: [256]u8 = undefined;
            const path = try std.fmt.bufPrint(&path_buf, "{s}{s}/{s}", .{ protocol.globals.SERVICE_SOCKET_DIR, namespace, name });
            const conn = try protocol.network.dial(io, path);

            self.io = io;
            self.namespace = namespace;
            self.name = name;
            self.conn = conn;
            self.is_connected = false;
            self.reader = conn.reader(io, &self.r_buf);
            self.writer = conn.writer(io, &self.w_buf);

            // BUGFIX-adjacent: a real spine-go service replies with *nothing
            // and closes the connection* on a type mismatch (see Service's
            // handshake comment) rather than a status byte — so a read that
            // errors out right after either handshake write means the same
            // thing here as an explicit non-OK status.
            try self.writer.interface.writeAll(mad.code(K));
            try self.writer.interface.flush();
            const key_status = self.reader.interface.takeInt(u8, .big) catch {
                conn.close(io);
                return ServiceCallError.PayloadTypeMismatch;
            };
            if (key_status != protocol.globals.OK_STATUS) {
                conn.close(io);
                return ServiceCallError.PayloadTypeMismatch;
            }

            try self.writer.interface.writeAll(mad.code(V));
            try self.writer.interface.flush();
            const value_status = self.reader.interface.takeInt(u8, .big) catch {
                conn.close(io);
                return ServiceCallError.PayloadTypeMismatch;
            };
            if (value_status != protocol.globals.OK_STATUS) {
                conn.close(io);
                return ServiceCallError.PayloadTypeMismatch;
            }

            self.is_connected = true;
            print("spine: connected to service '{s}'\n", .{name});
        }

        fn markDisconnected(self: *Self) void {
            self.is_connected = false;
            self.conn.close(self.io);
        }

        // BUGFIX: a mid-connection drop used to just surface as an error
        // forever after — nothing ever reconnected. Mirrors spine-go's
        // ServiceCaller.run(): if the last attempt marked the connection
        // dead, reconnect (unlimited backoff, same as the initial connect)
        // before trying this request. A failure *during* this request also
        // marks it dead so the *next* call reconnects first — this call
        // still returns its own error rather than silently retrying the same
        // request, since the request may have already taken effect
        // server-side by the time the failure is observed (unlike
        // Subscriber.next(), reading has no such idempotency concern).
        pub fn call(self: *Self, key: K) !V {
            try self.lock.lock(self.io);
            defer self.lock.unlock(self.io);

            if (!self.is_connected) {
                try self.connect(self.io, self.namespace, self.name);
            }

            const key_size = mad.getRequiredSize(K);
            var req_buf: [protocol.globals.MAX_PACKET_SIZE]u8 = undefined;
            _ = mad.encode(K, key, req_buf[0..key_size]);

            self.writer.interface.writeAll(req_buf[0..key_size]) catch |err| {
                self.markDisconnected();
                return err;
            };
            self.writer.interface.flush() catch |err| {
                self.markDisconnected();
                return err;
            };

            const value_size = mad.getRequiredSize(V);
            var resp_buf: [protocol.globals.MAX_PACKET_SIZE]u8 = undefined;
            var vecs: [1][]u8 = .{&resp_buf};
            const n = self.reader.interface.readVec(&vecs) catch |err| {
                self.markDisconnected();
                return err;
            };

            if (n == 0 or n < 1 + value_size) {
                self.markDisconnected();
                return ServiceCallError.CallFailed;
            }

            const status = resp_buf[0];
            if (status != protocol.globals.OK_STATUS) {
                print("spine: service '{s}' call failed with status {d}\n", .{ self.name, status });
                return ServiceCallError.CallFailed;
            }

            var value: V = undefined;
            _ = mad.decode(V, &value, resp_buf[1 .. 1 + value_size]);
            return value;
        }
    };
}

const testing = std.testing;
const Node = @import("node.zig").Node;
const test_support = @import("test_support.zig");
const testIo = test_support.testIo;
const testAllocator = test_support.testAllocator;

// ServiceCaller-centric behavior: handshake rejection and reconnect-on-drop.
// Handler dispatch behavior lives in service.zig instead.

fn doubleHandler(input: u32) anyerror!u32 {
    return input * 2;
}

test "service: mismatched key type is rejected, not retried forever" {
    const io = testIo();
    const allocator = testAllocator();

    var node = try Node.init("common", "service_test_keymismatch_node", io, allocator);
    defer node.deinit();

    _ = try node.newService(u32, u32, "service_test_keymismatch", doubleHandler);

    const result = node.newServiceCaller(f32, u32, "service_test_keymismatch");
    try testing.expectError(ServiceCallError.PayloadTypeMismatch, result);
}

test "service: mismatched value type is rejected" {
    const io = testIo();
    const allocator = testAllocator();

    var node = try Node.init("common", "service_test_valmismatch_node", io, allocator);
    defer node.deinit();

    _ = try node.newService(u32, u32, "service_test_valmismatch", doubleHandler);

    const result = node.newServiceCaller(u32, f32, "service_test_valmismatch");
    try testing.expectError(ServiceCallError.PayloadTypeMismatch, result);
}

test "service: caller reconnects when marked disconnected" {
    const io = testIo();
    const allocator = testAllocator();

    var node = try Node.init("common", "service_test_reconnect_node", io, allocator);
    defer node.deinit();

    _ = try node.newService(u32, u32, "service_test_reconnect", doubleHandler);
    const caller = try node.newServiceCaller(u32, u32, "service_test_reconnect");

    try testing.expectEqual(@as(u32, 2), try caller.call(1));

    // Simulate having detected a dead connection the way markDisconnected()
    // would after a real write/read failure, without actually touching the
    // fd: Service doesn't expose its accepted per-caller connections (unlike
    // Publisher's `clients`), so there's no equivalent of the pubsub
    // reconnect test's "close the peer's copy, not our own" trick available
    // here. This still exercises the exact branch that matters — call()
    // reconnecting before attempting the request when is_connected is false
    // — see readme.md's live cross-process verification for the real
    // socket-level failure path (killing and restarting an actual service).
    caller.is_connected = false;

    try testing.expectEqual(@as(u32, 6), try caller.call(3));
}

File: client-go/service_caller.go
package spine

import (
	"context"
	"fmt"
	"net"

	"github.com/cenkalti/backoff/v4"
	"github.com/poisnoir/spine-go/client-go/internal/globals"
	"github.com/poisnoir/spine-go/client-go/internal/mad"
)

type ServiceCaller[K any, V any] struct {
	node        *Node
	serviceName string

	keySerializer   *mad.Mad[K]
	valueSerializer *mad.Mad[V]

	ctx context.Context

	conn        net.Conn
	requests    chan serviceRequest[K, V]
	isConnected bool
}

func NewServiceCaller[K any, V any](node *Node, serviceName string) (*ServiceCaller[K, V], error) {

	keySer, err := mad.NewMad[K]()
	if err != nil {
		return nil, err
	}

	valueSer, err := mad.NewMad[V]()
	if err != nil {
		return nil, err
	}

	// Waiting for a service, not a publisher: isService = true.
	err = node.registerConsumer(serviceName, true)
	if err != nil {
		return nil, err
	}

	// BUGFIX: no more entity-level Close()/cancel — a service caller shares its
	// node's context directly and lives as long as the node does, same as
	// Subscriber/Publisher/Service.
	sc := &ServiceCaller[K, V]{
		node:        node,
		serviceName: serviceName,

		keySerializer:   keySer,
		valueSerializer: valueSer,

		ctx: node.ctx,

		isConnected: false,
		requests:    make(chan serviceRequest[K, V], 100),
	}

	bo := backoff.WithContext(backoff.NewExponentialBackOff(backoff.WithMaxElapsedTime(0)), sc.ctx)
	_ = backoff.Retry(sc.connect, bo)

	go sc.run()
	return sc, nil
}

func (sc *ServiceCaller[K, V]) run() {

	for {
		select {
		case <-sc.ctx.Done():
			if sc.conn != nil {
				sc.conn.Close()
			}
			return
		case requestData := <-sc.requests:
			if !sc.isConnected {
				bo := backoff.WithContext(backoff.NewExponentialBackOff(backoff.WithMaxElapsedTime(0)), sc.ctx)
				if err := backoff.Retry(sc.connect, bo); err != nil {
					requestData.output <- serviceOutput[V]{err: err}
					continue
				}
			}

			if err := requestData.ctx.Err(); err != nil {
				requestData.output <- serviceOutput[V]{err: err}
				continue
			}

			output, err := sc.send(requestData.input)
			if err != nil {
				sc.node.logger.Error("failed to send request", "error", err)
				sc.isConnected = false
			}
			requestData.output <- serviceOutput[V]{data: output, err: err}
		}
	}
}

// send is the gateway to kcp connection. It acts as multiplexer
func (sc *ServiceCaller[K, V]) send(key K) (V, error) {
	var v V

	requestSize := sc.keySerializer.GetRequiredSize()
	if requestSize > globals.MAX_PACKET_SIZE {
		return v, fmt.Errorf(globals.ERROR_PAYLOAD_SIZE)
	}

	bufPtr := sc.node.bufferPool.Get().(*[]byte)
	defer sc.node.bufferPool.Put(bufPtr)
	buf := *bufPtr

	sc.keySerializer.Encode(&key, buf)

	// _, err := write(sc.conn, buf, requestSize, true)
	_, err := sc.conn.Write(buf[:requestSize])
	if err != nil {
		return v, err
	}

	_, err = sc.conn.Read(buf)
	if err != nil {
		return v, err
	}

	if buf[0] != globals.OK_STATUS_CODE {
		// error messages can't be decoded here: mad has no string/slice support,
		// so the service side only ever sends a bare status code, not a message.
		return v, fmt.Errorf("call error: status %d", buf[0])
	}

	_ = sc.valueSerializer.Decode(buf[1:], &v)
	return v, nil
}

// Call sends key to the service and returns V from service
// context is used for establishing connection
func (sc *ServiceCaller[K, V]) Call(key K, ctx context.Context) (V, error) {

	var zero V
	data := serviceRequest[K, V]{
		ctx:    ctx,
		input:  key,
		output: make(chan serviceOutput[V], 1),
	}

	select {
	case sc.requests <- data:
	case <-ctx.Done():
		return zero, ctx.Err()
	case <-sc.ctx.Done():
		return zero, sc.ctx.Err()
	}

	select {
	case <-ctx.Done():
		return zero, ctx.Err()
	case <-sc.ctx.Done():
		return zero, sc.ctx.Err()
	case output := <-data.output:
		return output.data, output.err
	}
}

func (sc *ServiceCaller[K, V]) connect() error {

	logger := sc.node.logger.With(
		sc.node.Name(),
		"service_caller",
		sc.serviceName,
		"connect",
	)

	// establishing connection
	conn, err := net.Dial("unix", "/tmp/spine/service/"+sc.node.namespace+"/"+sc.serviceName)
	if err != nil {
		logger.Error("failed to dial service", "error", err)
		return err
	}

	// getting buffer for comm
	bufPtr := sc.node.bufferPool.Get().(*[]byte)
	defer sc.node.bufferPool.Put(bufPtr)
	buf := *bufPtr

	// validating input/output service types
	keyCode := sc.keySerializer.Code()
	n := copy(buf, keyCode)

	_, err = conn.Write(buf[:n])
	if err != nil {
		logger.Error("failed to write into socket", "error", err)
		return err
	}

	_, err = conn.Read(buf)
	if err != nil {
		logger.Error("failed to read from socket", "error", err)
	} else if buf[0] != globals.OK_STATUS_CODE {
		err = fmt.Errorf("service data type is different")
		logger.Error("failed to validate service input type", "error", err)
		return err
	}

	valueCode := sc.valueSerializer.Code()
	n = copy(buf, valueCode)

	_, err = conn.Write(buf[:n])
	if err != nil {
		logger.Error("failed to write into socket", "error", err)
		return err
	}
	_, err = conn.Read(buf)
	if err != nil {
		logger.Error("failed to read from socket", "error", err)
		return err
	} else if buf[0] != globals.OK_STATUS_CODE {
		err = fmt.Errorf("service data type is different")
		logger.Error("failed to validate service output type", "error", err)
		return err
	}
	sc.conn = conn
	sc.isConnected = true

	return nil
}
